datalad_next/url_operations/ssh.py

```diff
@@ -160,49 +165,55 @@ def download(self,
-                cmd=f'{SshUrlOperations._stat_cmd}; {SshUrlOperations._cat_cmd}',
-            )
-            stream = props.pop('_stream')
-            expected_size = props['content-length']
-            dst_fp = sys.stdout.buffer if to_path is None \
-                else open(to_path, 'wb')
-            # Localize variable access to minimize overhead
-            dst_fp_write = dst_fp.write
-            # download can start
-            self._progress_report_start(
-                progress_id,
-                ('Download %s to %s', from_url, to_path),
-                'downloading',
-                expected_size,
-            )
-            for chunk in stream:
-                # write data
-                dst_fp_write(chunk)
-                # compute hash simultaneously
-                hasher.update(chunk)
-                self._progress_report_update(
-                    progress_id, ('Downloaded chunk',), len(chunk))
-            props.update(hasher.get_hexdigest())
-            return props
-        except CommandError as e:
-            if e.code == 244:
-                # this is the special code for a file-not-found
-                raise UrlOperationsResourceUnknown(from_url) from e
-            else:
-                # wrap this into the datalad-standard, but keep the
-                # original exception linked
-                raise UrlOperationsRemoteError(from_url, message=str(e)) from e
-        finally:
-            if dst_fp and to_path is not None:
-                dst_fp.close()
-            self._progress_report_stop(progress_id, ('Finished download',))
+            with iter_subproc(cmd) as stream:
+                props = self._get_props(from_url, stream)
+                expected_size = props['content-length']
+                # The stream might have changed due to not yet processed, but
+                # fetched data, that is now chained in front of it. Therefore we
+                # get the updated stream from the props
+                download_stream = props.pop('_stream')
+
+                dst_fp = sys.stdout.buffer \
+                    if to_path is None \
+                    else open(to_path, 'wb')
+
+                # Localize variable access to minimize overhead
+                dst_fp_write = dst_fp.write
+
+                # download can start
+                for chunk in side_effect(
+                        lambda chunk: self._progress_report_update(
+                            progress_id,
+                            ('Downloaded chunk',),
+                            len(chunk)
+                        ),
+                        download_stream,
+                        before=partial(
+                            self._progress_report_start,
+                            progress_id,
+                            ('Download %s to %s', from_url, to_path),
+                            'downloading',
+                            expected_size
+                        )
```

Comment:
Nice pattern!

It appears to me that we should add a `stream_progress` helper to the base class that provides this side-effect without needing the verbosity of the setup. We just need a `progress_id` the texts and an `expected_size` (if available).

This particular implementation here seems to be missing the end of the progress report.

Reply:
Thanka

Reply:
Moved progress report end into the `after`-function of  `side_effect` (which is executed in a `finally`-branch, so it will also run in case of errors). (Done in 06f3ed56805fd9d537eea0b681a39b2f6608937f)

Reply:
> ...
> It appears to me that we should add a `stream_progress` helper to the base class that provides this side-effect without needing the verbosity of the setup. We just need a `progress_id` the texts and an `expected_size` (if available).
> ...

Nice idea. Done in 585525007caf0fb70b8336af9d2f0d4802e62624